libraries/AP_MSP/msp.cpp

```diff
@@ -0,0 +1,296 @@
+#include <AP_Math/AP_Math.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include "msp.h"
+#include <stdint.h>
+
+#if  HAL_MSP_ENABLED
+
+uint8_t MSP::msp_serial_checksum_buf(uint8_t checksum, const uint8_t *data, int len)
+{
+    while (len-- > 0) {
+        checksum ^= *data++;
+    }
+    return checksum;
+}
+
+int MSP::msp_serial_send_frame(msp_port_t *msp, const uint8_t * hdr, int hdr_len, const uint8_t * data, int data_len, const uint8_t * crc, int crc_len)
+{
+    if (msp->uart == nullptr) {
+        return 0;
+    }
+
+    // We are allowed to send out the response if
+    //  a) TX buffer is completely empty (we are talking to well-behaving party that follows request-response scheduling;
+    //     this allows us to transmit jumbo frames bigger than TX buffer (serialWriteBuf will block, but for jumbo frames we don't care)
+    //  b) Response fits into TX buffer
+    const uint32_t total_frame_length = hdr_len + data_len + crc_len;
+
+    if (!msp->uart->tx_pending() && (msp->uart->txspace() < total_frame_length)) {
```

Comment:
Some of this condition seem inverted compared to the comments.

Reply:
why?